File: C++/ASP Advanced/Arrays/Prefix Sums/Leetcode-238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size(), 0);
        vector<int> right(nums.size() + 1, 1);

        for (int i = nums.size() - 1; i >= 0; i--){
            right[i] = right[i + 1] * nums[i];
        }

        int left = 1;

        for (int i = 0; i < nums.size(); i++){
            res[i] = left * right[i + 1];
            left *= nums[i];
        }

        return res;
    }
};
// The code defines a solution to the problem of finding the product of all elements in an array except for the element at the current index.
// It uses two auxiliary arrays: one to store the cumulative product of elements to the right of each index and another to compute the final result.
// The function iterates through the array twice: first to fill the right products, and then to compute the result using both left and right products.
// The final result is returned as a vector containing the products for each index, excluding the element at that index.

File: Python/ASP Advanced/Arrays/Prefix Sums/Leetcode-238.py
# https://leetcode.com/problems/product-of-array-except-self/description/

from typing import List

class Solution:

    def productExceptSelfOptimal(self, nums: List[int]) -> List[int]:
        res = [1]*len(nums)

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]

        prefix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= prefix
            prefix *= nums[i]

        return res
    
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1]*len(nums)

        carry = 1
        for i in range(len(nums)):
            if i > 0:
                res[i] = res[i - 1] * carry

            carry = nums[i]

        res1 = [1] * len(nums)
        carry = 1
        for i in range(len(nums) - 1, -1, -1):
            if i < len(nums) - 1:
                res1[i] = res1[i + 1] * carry

            carry = nums[i]

        for i in range(len(nums)):
            res[i] *= res1[i]

        return res

sol = Solution()
nums = [1, 2, 3, 4]
print(sol.productExceptSelf(nums))